cli/src/cluster_query.rs

```diff
@@ -820,14 +830,40 @@ pub fn process_cluster_version(rpc_client: &RpcClient, config: &CliConfig) -> Pr
     }
 }
 
-pub fn process_fees(rpc_client: &RpcClient, config: &CliConfig) -> ProcessResult {
-    let result = rpc_client.get_recent_blockhash_with_commitment(config.commitment)?;
-    let (recent_blockhash, fee_calculator, last_valid_slot) = result.value;
-    let fees = CliFees {
-        slot: result.context.slot,
-        blockhash: recent_blockhash.to_string(),
-        lamports_per_signature: fee_calculator.lamports_per_signature,
-        last_valid_slot,
+pub fn process_fees(
+    rpc_client: &RpcClient,
+    config: &CliConfig,
+    blockhash: Option<&Hash>,
+) -> ProcessResult {
+    let result = if let Some(blockhash) = blockhash {
+        let result = rpc_client
+            .get_fee_calculator_for_blockhash_with_commitment(blockhash, config.commitment)?;
+        if let Some(fee_calculator) = result.value {
+            Some(Response {
+                context: result.context,
+                value: (*blockhash, fee_calculator, 0),
+            })
+        } else {
+            None
+        }
+    } else {
+        Some(rpc_client.get_recent_blockhash_with_commitment(config.commitment)?)
+    };
+    let fees = if let Some(result) = result {
+        let (recent_blockhash, fee_calculator, last_valid_slot) = result.value;
+        let last_valid_slot = if last_valid_slot == 0 {
+            None
+        } else {
+            Some(last_valid_slot)
+        };
+        CliFees::some(
+            result.context.slot,
+            recent_blockhash,
+            fee_calculator.lamports_per_signature,
+            last_valid_slot,
+        )
+    } else {
+        CliFees::none()
```

Comment:
unifying the code looks nice; but considering these https://github.com/solana-labs/solana/pull/15555/files#r583700280 type system loophole, I guess, we might be better off creating CliFees::none() / ::some() at each branches in the preceding if branches?